test_util/launch.py

```diff
@@ -1,20 +1,257 @@
+            'private_agents': convert_host_list(cf.get_private_agent_ips()),
+            'public_agents': convert_host_list(cf.get_public_agent_ips())})
+        return desc
+
+    def delete(self, info):
+        if info['ssh']['delete_with_stack']:
+            self.boto_wrapper.delete_key_pair(info['ssh']['key_name'])
+        self.get_stack(info).delete()
+
+    def ssh_from_config(self, config):
+        """
+        AWS clusters will have KeyName parameter that will be required in the template
+        if this paramter is given, then assume that the key already exists. If this is absent,
+        then generate the key, amend the config, and then delete the key with the cluster
+        Accepts a YAML like:
+        ---
+        ssh_info:
+          - user: user to ssh into cluster with.
+          - home_dir: path on remote host where files can be written
+          - private_key_path: path to the private ssh key (if key is not generated)
+        """
+        generate_key = False
+        key_name = None
+        private_key = None
+        # Native AWS parameters take precedence
+        if 'parameters' in config:
+            for p in config['parameters']:
+                if p['ParameterKey'] == 'KeyName':
+                    key_name = p['ParameterValue']
+        # check ssh_info data
+        if 'ssh_info' in config:
+            if key_name is None:
+                key_name = config['ssh_info'].get('key_name', None)
+            elif 'key_name' in config['ssh_info']:
+                raise LauncherError('ValidationError',
+                                    'Provide key name as either a parameter or in ssh_info; not both')
+            if 'private_key_path' in config['ssh_info']:
+                if key_name is None:
+                    raise LauncherError('ValidationError', 'If a private_key_path is provided, '
+                                        'then the KeyName template parameter must be set')
+                private_key = load_string(config['ssh_info']['private_key_path'])
+        if not key_name:
+            key_name = config['stack_name']
+            generate_key = True
+            private_key = self.boto_wrapper.create_key_pair(key_name)
+            cf_params = config.get('parameters', list())
+            cf_params.append({'ParameterKey': 'KeyName', 'ParameterValue': key_name})
+            config['parameteres'] = cf_params
+        return {
+            'delete_with_stack': generate_key,
+            'private_key': private_key,
+            'key_name': key_name,
+            'user': config.get('ssh_info', dict()).get('user', None)}
+
+    def test(self, info, test_cmd):
+        ssh_info = info['ssh']
+        for param in ['user', 'private_key']:
+            if param not in ssh_info:
+                raise LauncherError('MissingInfo', 'Cluster was created without providing ssh_info: {}'.format(param))
+        details = self.describe(info)
+        # FIX ME clean this up when done
```

Comment:
Possibly. What I intended here was "TODO: generalize this" so I should at least make that change.